File: shapeapp-manhood_iOS/Histo/Classes/Model/UserData/UsersData.m
//
//  UsersData.m
//  CoolHisto
//
//  Created by Viktor Gubriienko on 24.03.14.
//  Copyright (c) 2014 Viktor Gubriienko. All rights reserved.
//

#import "UsersData.h"

static NSString *const kJSONLengthKey               = @"length";
static NSString *const kJSONGirthKey                = @"girth";
static NSString *const kJSONThicknessKey            = @"thickness";
static NSString *const kJSONLookupUsersKey          = @"lookup_users";
static NSString *const kJSONBinKey                  = @"bin";
static NSString *const kJSONCountKey                = @"count";
static NSString *const kJSONAverageKey              = @"average";
static NSInteger const kJSONUserLengthIndex         = 0;
static NSInteger const kJSONUserGirthIndex          = 1;
static NSInteger const kJSONUserThicknessIndex      = 2;
static NSInteger const kJSONUserCirclesIndex        = 3;
static NSInteger const kJSONUserLengthPosIndex      = 4;
static NSInteger const kJSONUserGirthPosIndex       = 5;
static NSInteger const kJSONUserThicknessPosIndex   = 6;
static NSInteger const kJSONUserCircleBaseIndex     = 0;
static NSInteger const kJSONUserCircleMidIndex      = 1;
static NSInteger const kJSONUserCircleUpperIndex    = 2;
static NSInteger const kJSONUserCircleTipIndex      = 3;
static NSInteger const kJSONUserCircleXIndex        = 0;
static NSInteger const kJSONUserCircleYIndex        = 1;
static NSInteger const kSearchValueIndex            = 0;
static NSInteger const kSearchUserIDIndex           = 1;


@implementation UsersData {
    NSDictionary *_usersData;
    NSDictionary *_searchData;
}

- (instancetype)initWithContentsOfFile:(NSString*)filePath rootKey:(NSString*)rootKey {
    self = [super init];
    if (self) {
        NSData *jsonData = [NSData dataWithContentsOfFile:filePath];
        if ( jsonData ) {
            _usersData = [NSPropertyListSerialization propertyListWithData:jsonData
                                                                   options:kCFPropertyListMutableContainersAndLeaves
                                                                    format:nil
                                                                     error:nil];
            [self setupSearchData];            
        } else {
            return self = nil;
        }
        
    }
    return self;
    
}

#pragma mark - DynamicProperties

- (void)setupSearchData {
    _searchData = @{kJSONLengthKey: [self createSearchDataForUserDataIndex:kJSONUserLengthIndex positionValueIndex:kJSONUserLengthPosIndex],
                    kJSONGirthKey: [self createSearchDataForUserDataIndex:kJSONUserGirthIndex positionValueIndex:kJSONUserGirthPosIndex],
                    kJSONThicknessKey: [self createSearchDataForUserDataIndex:kJSONUserThicknessIndex positionValueIndex:kJSONUserThicknessPosIndex]};
}

- (NSArray*)createSearchDataForUserDataIndex:(NSInteger)index positionValueIndex:(NSInteger)positionIndex {
    NSMutableArray *searchData = [[NSMutableArray alloc] initWithCapacity:[_usersData[kJSONLookupUsersKey] count]];
    for (NSString *userID in _usersData[kJSONLookupUsersKey]) {
        NSArray *userData = _usersData[kJSONLookupUsersKey][userID];
        [searchData addObject:@[userData[index], userID]];
    }
    
    [searchData sortUsingComparator:^NSComparisonResult(NSArray *userData1, NSArray *userData2) {
        return [userData1[kSearchValueIndex] compare:userData2[kSearchValueIndex]];
    }];
    
    [searchData enumerateObjectsUsingBlock:^(NSArray *searchUserData, NSUInteger idx, BOOL *stop) {
        [_usersData[kJSONLookupUsersKey][searchUserData[kSearchUserIDIndex]] insertObject:@(idx) atIndex:positionIndex];
    }];
    
    return [searchData copy];
}

- (NSArray*)searchUserDataForValue:(float)value withKey:(NSString*)key exact:(BOOL)exact {

    NSArray *allData = _searchData[key];
    NSArray *result;
    
    if ( exact ) {
        
        if ( value < [allData.firstObject[kSearchValueIndex] floatValue] || value > [allData.lastObject[kSearchValueIndex] floatValue] ) {
            result = nil;
        } else {
            for (NSArray *userData in allData) {
                if ( [userData[kSearchValueIndex] floatValue] == value ) {
                    result = userData;
                    break;
                }
            }
        }
        
    } else {
        float valueDiff = MAXFLOAT;
        
        for (NSArray *userData in allData) {
            float diff = fabsf([userData[kSearchValueIndex] floatValue] - value);
            if ( diff < valueDiff ) {
                result = userData;
                valueDiff = diff;
            }
        }

    }
    
    return result;

}

#pragma mark - Public

- (float)lengthOfUserWithID:(NSString*)userID {
    return [_usersData[kJSONLookupUsersKey][userID][kJSONUserLengthIndex] floatValue];
}

- (float)girthOfUserWithID:(NSString*)userID {
    return [_usersData[kJSONLookupUsersKey][userID][kJSONUserGirthIndex] floatValue];
}

- (float)thicknessOfUserWithID:(NSString*)userID {
    return [_usersData[kJSONLookupUsersKey][userID][kJSONUserThicknessIndex] floatValue];
}

- (CGPoint)basePositionOfUserWithID:(NSString*)userID {
    return CGPointMake([_usersData[kJSONLookupUsersKey][userID][kJSONUserCirclesIndex][kJSONUserCircleBaseIndex][kJSONUserCircleXIndex] floatValue],
                       [_usersData[kJSONLookupUsersKey][userID][kJSONUserCirclesIndex][kJSONUserCircleBaseIndex][kJSONUserCircleYIndex] floatValue]);
}

- (CGPoint)midPositionOfUserWithID:(NSString*)userID {
    return CGPointMake([_usersData[kJSONLookupUsersKey][userID][kJSONUserCirclesIndex][kJSONUserCircleMidIndex][kJSONUserCircleXIndex] floatValue],
                       [_usersData[kJSONLookupUsersKey][userID][kJSONUserCirclesIndex][kJSONUserCircleMidIndex][kJSONUserCircleYIndex] floatValue]);
}

- (CGPoint)upperPositionOfUserWithID:(NSString*)userID {
    return CGPointMake([_usersData[kJSONLookupUsersKey][userID][kJSONUserCirclesIndex][kJSONUserCircleUpperIndex][kJSONUserCircleXIndex] floatValue],
                       [_usersData[kJSONLookupUsersKey][userID][kJSONUserCirclesIndex][kJSONUserCircleUpperIndex][kJSONUserCircleYIndex] floatValue]);
}

- (CGPoint)tipPositionOfUserWithID:(NSString*)userID {
    return CGPointMake([_usersData[kJSONLookupUsersKey][userID][kJSONUserCirclesIndex][kJSONUserCircleTipIndex][kJSONUserCircleXIndex] floatValue],
                       [_usersData[kJSONLookupUsersKey][userID][kJSONUserCirclesIndex][kJSONUserCircleTipIndex][kJSONUserCircleYIndex] floatValue]);
}

- (NSString*)userIDWithNearestLength:(float)length {
    return [self searchUserDataForValue:length withKey:kJSONLengthKey exact:NO][kSearchUserIDIndex];
}

- (NSString*)userIDWithNearestGirth:(float)girth {
    return [self searchUserDataForValue:girth withKey:kJSONGirthKey exact:NO][kSearchUserIDIndex];
}

- (NSString*)userIDWithNearestThickness:(float)thickness {
    return [self searchUserDataForValue:thickness withKey:kJSONThicknessKey exact:NO][kSearchUserIDIndex];
}

- (NSUInteger)positionOfUserWithNearestLength:(float)length {
    return [_searchData[kJSONLengthKey] indexOfObject:[self searchUserDataForValue:length withKey:kJSONLengthKey exact:NO]];
}

- (NSUInteger)positionOfUserWithNearestGirth:(float)girth {
    return [_searchData[kJSONGirthKey] indexOfObject:[self searchUserDataForValue:girth withKey:kJSONGirthKey exact:NO]];
}

- (NSUInteger)positionOfUserWithNearestThickness:(float)thickness {
    return [_searchData[kJSONThicknessKey] indexOfObject:[self searchUserDataForValue:thickness withKey:kJSONThicknessKey exact:NO]];
}

- (NSString*)userIDByLengthPosition:(NSUInteger)position {
    if (position >= self.usersCount) {
        return nil;
    } else {
        return _searchData[kJSONLengthKey][position][kSearchUserIDIndex];
    }
}

- (NSString*)userIDByGirthPosition:(NSUInteger)position {
    if (position >= self.usersCount) {
        return nil;
    } else {
        return _searchData[kJSONGirthKey][position][kSearchUserIDIndex];
    }
}

- (NSString*)userIDByThicknessPosition:(NSUInteger)position {
    if (position >= self.usersCount) {
        return nil;
    } else {
        return _searchData[kJSONThicknessKey][position][kSearchUserIDIndex];
    }
}

- (NSUInteger)positionByLengthOfUserWithID:(NSString*)userID {
    return [_usersData[kJSONLookupUsersKey][userID][kJSONUserLengthPosIndex] unsignedIntegerValue];
}

- (NSUInteger)positionByGirthOfUserWithID:(NSString*)userID {
    return [_usersData[kJSONLookupUsersKey][userID][kJSONUserGirthPosIndex] unsignedIntegerValue];
}

- (NSUInteger)positionByThicknessOfUserWithID:(NSString*)userID {
    return [_usersData[kJSONLookupUsersKey][userID][kJSONUserThicknessPosIndex] unsignedIntegerValue];
}

#pragma mark - DynamicProperties

- (NSInteger)usersCount {
    return [_usersData[kJSONLookupUsersKey] count];
}

- (NSArray *)lengthBins {
    return _usersData[kJSONLengthKey][kJSONBinKey];
}

- (NSArray *)lengthCounts{
    return _usersData[kJSONLengthKey][kJSONCountKey];
}

- (NSArray *)girthBins {
    return _usersData[kJSONGirthKey][kJSONBinKey];
}

- (NSArray *)girthCounts {
    return _usersData[kJSONGirthKey][kJSONCountKey];
}

- (NSArray *)thicknessBins {
    return _usersData[kJSONThicknessKey][kJSONBinKey];
}

- (NSArray *)thicknessCounts {
    return _usersData[kJSONThicknessKey][kJSONCountKey];
}

- (float)averageLength {
    return [_usersData[kJSONAverageKey][kJSONUserLengthIndex] floatValue];
}

- (float)averageGirth {
    return [_usersData[kJSONAverageKey][kJSONUserGirthIndex] floatValue];
}

- (float)averageThickness {
    return [_usersData[kJSONAverageKey][kJSONUserThicknessIndex] floatValue];
}

@end

File: shapeapp-manhood_android/ManHood/src/com/VVTeam/ManHood/DataModel/UsersData.java
package com.VVTeam.ManHood.DataModel;

import android.graphics.PointF;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;

public class UsersData {
	
	float MAXFLOAT = 0.05f; //0x1.fffffep+127f;
	
	ArrayList<Float> lengthBins;
	ArrayList<Float> lengthCounts;
	ArrayList<Float> girthBins;
	ArrayList<Float> girthCounts;
	ArrayList<Float> thicknessBins;
	ArrayList<Float> thicknessCounts;
	float averageLength;
	float averageGirth;
	float averageThickness;

	int usersCount;
	
    private JSONObject _usersData;
    private JSONObject _searchData;	


    
    public static JSONArray sort(JSONArray array, Comparator c){
        List<Object>    asList = new ArrayList(array.length());
        for (int i=0; i<array.length(); i++){
          asList.add(array.opt(i));
        }
        Collections.sort(asList, c);
        JSONArray  res = new JSONArray();
        for (Object o : asList){
          res.put(o);
        }
        return res;
    }
    
    
    public UsersData(JSONObject jsonObj) {
    	
    	_usersData = jsonObj;
    	
    	setupSearchData();
    }
    
    public void setupSearchData()
    {
    	_searchData = new JSONObject();
    	
    	try {
    		
			_searchData.put(JSONIndex.kJSONLengthKey, createSearchDataForUserDataIndex(JSONIndex.kJSONUserLengthIndex, JSONIndex.kJSONUserLengthPosIndex));
	    	_searchData.put(JSONIndex.kJSONGirthKey, createSearchDataForUserDataIndex(JSONIndex.kJSONUserGirthIndex, JSONIndex.kJSONUserGirthPosIndex));
	    	_searchData.put(JSONIndex.kJSONThicknessKey, createSearchDataForUserDataIndex(JSONIndex.kJSONUserThicknessIndex, JSONIndex.kJSONUserThicknessPosIndex));
		
    	} catch (JSONException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}

    }
    
    class SearchField {
    	float value = 0.0f;
    	String userId = "";
    	
    	public SearchField(float value, String userId) {
			// TODO Auto-generated constructor stub
    		this.value = value;
    		this.userId = userId;    				
		}
    }
    
    @SuppressWarnings("unchecked")
    public JSONArray createSearchDataForUserDataIndex(int index, int positionIndex) 
    {
    	try {
    		JSONArray searchData = new JSONArray();
//    		List<SearchField> searchList = new ArrayList<UsersData.SearchField>();
    		
    		
    		JSONObject objLookUp = _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey);
    		
    		Iterator<String> iterator = objLookUp.keys();
    		while(iterator.hasNext()){
    		    String userID = (String)iterator.next();
    		    
    		    JSONArray userData = objLookUp.getJSONArray(userID);

    		    JSONArray putObj = new JSONArray();
    		    
    		    putObj.put(JSONIndex.kSearchValueIndex, userData.get(index));
    		    putObj.put(JSONIndex.kSearchUserIDIndex, userID);
    		    
    		    searchData.put(putObj);
    		    
//    		    searchList.add(new SearchField(Float.parseFloat(userData.get(index).toString()), userID));
    		    
    		}
    	
//    		Collections.sort(searchList, new Comparator<SearchField>() {
//
//				@Override
//				public int compare(SearchField lhs, SearchField rhs) {
//					// TODO Auto-generated method stub
//					
//					return Float.compare(lhs.value, rhs.value);
//					
//				}
//			});
    		
//    		for (SearchField item : searchList) {
//    			 JSONArray putObj = new JSONArray();
//     		    
//     		    putObj.put(JSONIndex.kSearchValueIndex, item.value);
//     		    putObj.put(JSONIndex.kSearchUserIDIndex, item.userId);
//     		    
//     		    searchData.put(putObj);
//    		}
    		searchData = sort(searchData, new Comparator(){
    			   public int compare(Object a, Object b){
    			      JSONArray    ja = (JSONArray)a;
    			      JSONArray    jb = (JSONArray)b;
    			      
    			      try {
    			    	  float fa = Float.parseFloat(ja.get(JSONIndex.kSearchValueIndex).toString());
    			    	  float fb = Float.parseFloat(jb.get(JSONIndex.kSearchValueIndex).toString());
    			    	  return Float.compare(fa, fb);
    			      } catch (NumberFormatException e) {
						// TODO Auto-generated catch block
						e.printStackTrace();
    			      } catch (JSONException e) {
						// TODO Auto-generated catch block
						e.printStackTrace();
    			      }
    			      return 0;
    			   }
    			});
    			      
    		for (int i = 0 ; i < searchData.length(); i ++) {
    			String userId = searchData.getJSONArray(i).getString(JSONIndex.kSearchUserIDIndex);
    			
    			_usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userId).put(positionIndex, i);
    			
    		}
    		
    		return searchData;
    		
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}

    	return null;
    }
    
    public JSONArray searchUserDataForValue(float value, String key, boolean exact) {

    	JSONArray result = null;

    	try {
    	JSONArray allData = _searchData.getJSONArray(key);
        
        if ( exact ) {
            
        	JSONArray firstObj = allData.getJSONArray(0);
        	JSONArray lastObj = allData.getJSONArray(allData.length() - 1);
        	
            if ( value < firstObj.getDouble(JSONIndex.kSearchValueIndex) 
            		|| value > lastObj.getDouble(JSONIndex.kSearchValueIndex) ) {
                result = null;
            } else {
            	for (int i = 0 ; i < allData.length() ; i ++) {
            		JSONArray userData = allData.getJSONArray(i);
            		
            		if (userData.getDouble(JSONIndex.kSearchValueIndex) == value) {
            			result = userData;
            			break;
            		}
            	}
            }
            
        } else {
            float valueDiff = MAXFLOAT;
            
//            System.out.println("value = " + value);
//            System.out.println("alldata = " + allData.toString());
            
            for (int i = 0 ; i < allData.length() ; i ++) {
        		JSONArray userData = allData.getJSONArray(i);
        		float diff = (float) Math.abs((float)userData.getDouble(JSONIndex.kSearchValueIndex) - value);
        		if ( diff < valueDiff ) {
                    result = userData;
                    valueDiff = diff;
                }
        	}

        }
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	
        return result;

    }

    public float lengthOfUserWithID(String userID) {
    	try {
    		return Float.parseFloat(_usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getString(JSONIndex.kJSONUserLengthIndex).toString());
    	}
    	catch (JSONException e) {
    		e.printStackTrace();
    	}
    	catch (Exception e) {
			// TODO: handle exception
    		e.printStackTrace();
		}
    	return 0.0f;
    }

    public float girthOfUserWithID(String userID) {
    	try {
    		return Float.parseFloat(_usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getString(JSONIndex.kJSONUserGirthIndex).toString());
    	}
    	catch (JSONException e) {
    		e.printStackTrace();
    	}
    	catch (Exception e) {
			// TODO: handle exception
    		e.printStackTrace();
		}
    	return 0.0f;
    }

    public float thicknessOfUserWithID(String userID) {
    	try {
    		return Float.parseFloat(_usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getString(JSONIndex.kJSONUserThicknessIndex).toString());
    	}
    	catch (JSONException e) {
    		e.printStackTrace();
    	}
    	catch (Exception e) {
			// TODO: handle exception
    		e.printStackTrace();
		}
    	return 0.0f;
    }

    public PointF basePositionOfUserWithID(String userID) {
    	try {
    		return new PointF((float) _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getJSONArray(JSONIndex.kJSONUserCirclesIndex).getJSONArray(JSONIndex.kJSONUserCircleBaseIndex).getDouble(JSONIndex.kJSONUserCircleXIndex),
    				(float) _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getJSONArray(JSONIndex.kJSONUserCirclesIndex).getJSONArray(JSONIndex.kJSONUserCircleBaseIndex).getDouble(JSONIndex.kJSONUserCircleYIndex));
    	}
    	catch (JSONException e) {
    		e.printStackTrace();
    	}
    	return new PointF(0.0f, 0.0f);
//    	return null;
    }
    
    public PointF midPositionOfUserWithID(String userID) {
    	try {
    		return new PointF((float) _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getJSONArray(JSONIndex.kJSONUserCirclesIndex).getJSONArray(JSONIndex.kJSONUserCircleMidIndex).getDouble(JSONIndex.kJSONUserCircleXIndex),
    						  (float) _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getJSONArray(JSONIndex.kJSONUserCirclesIndex).getJSONArray(JSONIndex.kJSONUserCircleMidIndex).getDouble(JSONIndex.kJSONUserCircleYIndex));
    	}
    	catch (JSONException e) {
    		e.printStackTrace();
    	}
    	return new PointF(0.0f, 0.0f);
//    	return null;
    }

    public PointF upperPositionOfUserWithID(String userID) {
    	try {
    		return new PointF((float) _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getJSONArray(JSONIndex.kJSONUserCirclesIndex).getJSONArray(JSONIndex.kJSONUserCircleUpperIndex).getDouble(JSONIndex.kJSONUserCircleXIndex),
    						  (float) _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getJSONArray(JSONIndex.kJSONUserCirclesIndex).getJSONArray(JSONIndex.kJSONUserCircleUpperIndex).getDouble(JSONIndex.kJSONUserCircleYIndex));
    	}
    	catch (JSONException e) {
    		e.printStackTrace();
    	}
    	return new PointF(0.0f, 0.0f);
//    	return null;
    }
    
    public PointF tipPositionOfUserWithID(String userID) {
    	try {
    		return new PointF((float) _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getJSONArray(JSONIndex.kJSONUserCirclesIndex).getJSONArray(JSONIndex.kJSONUserCircleTipIndex).getDouble(JSONIndex.kJSONUserCircleXIndex),
    						  (float) _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getJSONArray(JSONIndex.kJSONUserCirclesIndex).getJSONArray(JSONIndex.kJSONUserCircleTipIndex).getDouble(JSONIndex.kJSONUserCircleYIndex));
    	}
    	catch (JSONException e) {
    		e.printStackTrace();
    	}
    	return new PointF(0.0f, 0.0f);
//    	return null;
    }
    
    public String userIDWithNearestLength(float length) {
    	try {
    		return searchUserDataForValue(length, JSONIndex.kJSONLengthKey, false).getString(JSONIndex.kSearchUserIDIndex).toString();
    	} catch (JSONException e) {
    		e.printStackTrace();
    	} catch (Exception e) {
			// TODO: handle exception
		}
    	
    	return "";
    }
    public String userIDWithNearestGirth(float girth) {
    	try {
    		return searchUserDataForValue(girth, JSONIndex.kJSONGirthKey, false).getString(JSONIndex.kSearchUserIDIndex).toString();
    	} catch (JSONException e) {
    		e.printStackTrace();
    	} catch (Exception e) {
			// TODO: handle exception
		}
    	return "";
    }
    public String userIDWithNearestThickness(float thickness) {
    	try {
    		return searchUserDataForValue(thickness, JSONIndex.kJSONThicknessKey, false).getString(JSONIndex.kSearchUserIDIndex).toString();
    	} catch (JSONException e) {
    		e.printStackTrace();
    	} catch (Exception e) {
			// TODO: handle exception
		}
    	return "";
    }

    public int positionOfUserWithNearestLength(float length) {
    	try {
    		JSONArray arrayValue = searchUserDataForValue(length, JSONIndex.kJSONLengthKey, false);
        	if (arrayValue != null) {
        		JSONArray array = _searchData.getJSONArray(JSONIndex.kJSONLengthKey);
        		for (int i = 0; i < array.length() ; i ++) {
        			if (array.getJSONArray(i) == arrayValue) {
        				return i;
        			}
        		}
        	}
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	
    	return -1;
    }
    
    public int positionOfUserWithNearestGirth(float girth) {
    	try {
    		JSONArray arrayValue = searchUserDataForValue(girth, JSONIndex.kJSONGirthKey, false);
        	if (arrayValue != null) {
        		JSONArray array = _searchData.getJSONArray(JSONIndex.kJSONGirthKey);
        		for (int i = 0; i < array.length() ; i ++) {
        			if (array.getJSONArray(i) == arrayValue) {
        				return i;
        			}
        		}
        	}
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	return -1;
    }
    
    public int positionOfUserWithNearestThickness(float thickness) {
    	try {
    		JSONArray arrayValue = searchUserDataForValue(thickness, JSONIndex.kJSONThicknessKey, false);
        	if (arrayValue != null) {
        		JSONArray array = _searchData.getJSONArray(JSONIndex.kJSONThicknessKey);
        		for (int i = 0; i < array.length() ; i ++) {
        			if (array.getJSONArray(i) == arrayValue) {
        				return i;
        			}
        		}
        	}
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	return -1;
    }

    public int userIDByLengthPosition(int position) {
        if (position >= usersCount) {
            return -1;
        } else {
        	try {
        		return _searchData.getJSONArray(JSONIndex.kJSONLengthKey).getJSONArray(position).getInt(JSONIndex.kSearchUserIDIndex);
        	} catch (JSONException e) {
        		e.printStackTrace();
        	}
        	return -1;
            
        }
    }
    
    public int userIDByGirthPosition(int position) {
        if (position >= usersCount) {
            return -1;
        } else {
        	try {
        		return _searchData.getJSONArray(JSONIndex.kJSONGirthKey).getJSONArray(position).getInt(JSONIndex.kSearchUserIDIndex);
        	} catch (JSONException e) {
        		e.printStackTrace();
        	}
        	return -1;
            
        }
    }
    
    public int userIDByThicknessPosition(int position) {
        if (position >= usersCount) {
            return -1;
        } else {
        	try {
        		return _searchData.getJSONArray(JSONIndex.kJSONThicknessKey).getJSONArray(position).getInt(JSONIndex.kSearchUserIDIndex);
        	} catch (JSONException e) {
        		e.printStackTrace();
        	}
        	return -1;
            
        }
    }
    
    public int positionByLengthOfUserWithID(String userID) {
    	try {
			
    		Object obj = _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).get(JSONIndex.kJSONUserLengthPosIndex);
    		
    		return Integer.parseInt(obj.toString());
    		
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	
    	return -1;
    }
    public int positionByGirthOfUserWithID(String userID) {
    	try {
    		
    		return _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getInt(JSONIndex.kJSONUserGirthPosIndex);
    		
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	
    	return -1;
    }
    public int positionByThicknessOfUserWithID(String userID) {
    	try {
    		
    		return _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).getJSONArray(userID).getInt(JSONIndex.kJSONUserThicknessPosIndex);
    		
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	
    	return -1;
    }
    
    @SuppressWarnings("unchecked")
    public int getUsersCount() {

    	int count = 0;

    	try {
        	
			Iterator<String> keys = _usersData.getJSONObject(JSONIndex.kJSONLookupUsersKey).keys();
        	
        	while (keys.hasNext()) {
        		String key = keys.next();
        		
        		count ++;        	    
        	}
        	
    	} catch (JSONException e) {
    		e.printStackTrace();
    	} catch (Exception e) {
    		e.printStackTrace();
		}
    	
    	return count;
    }
    public JSONArray getlengthBins() {
    	
    	try {
    		return _usersData.getJSONObject(JSONIndex.kJSONLengthKey).getJSONArray(JSONIndex.kJSONBinKey);
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	
    	return null;
    }
    public JSONArray getlengthCounts() {
    	try {
    		return _usersData.getJSONObject(JSONIndex.kJSONLengthKey).getJSONArray(JSONIndex.kJSONCountKey);
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	
    	return null;
    }
    public JSONArray getgirthBins() {
    	
    	try {
    		return _usersData.getJSONObject(JSONIndex.kJSONGirthKey).getJSONArray(JSONIndex.kJSONBinKey);
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	
    	return null;
    }
    public JSONArray getgirthCounts() {
    	try {
    		return _usersData.getJSONObject(JSONIndex.kJSONGirthKey).getJSONArray(JSONIndex.kJSONCountKey);
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	
    	return null;
    }
    public JSONArray getthicknessBins() {
    	
    	try {
    		return _usersData.getJSONObject(JSONIndex.kJSONThicknessKey).getJSONArray(JSONIndex.kJSONBinKey);
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	
    	return null;
    }
    public JSONArray getthicknessCounts() {
    	try {
    		return _usersData.getJSONObject(JSONIndex.kJSONThicknessKey).getJSONArray(JSONIndex.kJSONCountKey);
    	} catch (JSONException e) {
    		e.printStackTrace();
    	}
    	
    	return null;
    }
    public float getaverageLength() {
    	try {
    		JSONArray jsonObj = _usersData.getJSONArray(JSONIndex.kJSONAverageKey);
    		
    		return (float) jsonObj.getDouble(JSONIndex.kJSONUserLengthIndex);
    	    
    	} catch (JSONException e) {
    		e.printStackTrace();
    	} catch (Exception e) {

		}
    	
    	return 0.0f;
    }
    public float getaverageGirth() {
    	try {
    		JSONArray jsonObj = _usersData.getJSONArray(JSONIndex.kJSONAverageKey);
    		
    		return (float) jsonObj.getDouble(JSONIndex.kJSONUserGirthIndex);
    	    
    	} catch (JSONException e) {
    		e.printStackTrace();
    	} catch (Exception e) {

		}
    	
    	return 0.0f;
    }
    public float getaverageThickness() {
    	try {
    		JSONArray jsonObj = _usersData.getJSONArray(JSONIndex.kJSONAverageKey);
    		
    		return (float) jsonObj.getDouble(JSONIndex.kJSONUserThicknessIndex);
    	    
    	} catch (JSONException e) {
    		e.printStackTrace();
    	} catch (Exception e) {

		}
    	
    	return 0.0f;
    }
}
